### ISSUE TITLE
prepped format should not include toc, authors section, and index

### ISSUE THREAD

**Issue by reschke** [ietf-tools/xml2rfc#791]
All of these can easily be re-generated.

Furthermore:

- inclusion of "authors" section leaks into author-visible grammar (author as child element of section)
- auto-generated appendices would need to be distinguishable from real contents, so that the processor can decide how to deal with it
- using "pn" IDs as targets is xref elements is invalid according to the spec 

(and there are probably many more reasons; feel free to add :-).

**Comment by jrlevine** [comment]
It turns out that xml2rfc uses the `<toc>` section when rendering RFCs, so I think we're stuck with it.

**Comment by reschke** [comment]
That's fixable. You know, in absence of \<toc>, it will be happy to generate it.

**Comment by reschke** [comment]
Well, Robert is wrong. If we ever change the algorithm that generates the TOC, we *want* the published specs (be it XML or HTML) to reflect that change.

That said; the main issues are preptool extensions leaking into the authoring grammar, and auto-generated sections that need heuristics to be distinguisable from authored content.

**Comment by JayDaley** [comment]
I have so many questions:

1. Why would we not have a "guarantee that a regenerated <toc> would be the same as the one that's there" - isn't it the same tool?
2. Is there any reason not to document the **pn** attribute of the \<t\> element and explain to people how it works but that it will be completely rewritten by the prep tool if they choose to use it?  Would it matter if someone used it in an I-D and submitted that?
3. What's wrong with "auto-generated sections that need heuristics to be distinguishable from authored content" if those sections are only related to document structure and not to the actual content?  i.e. we're in the content business not the formatting business.
4. Should we instead address all of this by being much clearer that authoring RFCXML is not the same as publishing RFCXML and maybe differentiate them formally with two distinct but related names such as RFCXML and RFCXML/A to mirror the PDF naming, and then document the additional grammar supported in RFCXML/A (or RFCXML/PREPPED or whatever)?

**Comment by reschke** [comment]
> Is there any reason not to document the pn attribute of the <t> element and explain to people how it works but that it will be completely rewritten by the prep tool if they choose to use it? Would it matter if someone used it in an I-D and submitted that?

Adding things to the author visible vocabulary that are actually not for use by the authors is a disservice to authors.

> What's wrong with "auto-generated sections that need heuristics to be distinguishable from authored content" if those sections are only related to document structure and not to the actual content? i.e. we're in the content business not the formatting business.

It means that any tool the processes rfcxml input needs to hardwire the same heuristics to skip this content. ("Oh, a back section containing "author" tags - let's skip that")

> Should we instead address all of this by being much clearer that authoring RFCXML is not the same as publishing RFCXML

Absolutely. But that additional grammar actually needs to be well-defined. For instance, the published "canonical" XML for RFC 9114 currrently is invalid, in that it contains a back section containg an index with xref's pointing to "pn" attributes. Any processor written according to the specs will trip over this.

**Comment by rjsparks** [comment]
What Julian is missing in my argument is that _if_ we change something like how we generate ToC, we still want to know what was generated at the time the RFC was published. (No, I _don't_ think that we would necessarily re-render older RFCs with a newer way of ToC-ing should this ever happen.) This keeps truth in the right place.

I would be fine if there were a way to say things like this that didn't involve bleeding into the authoring grammar, but that would require a different _publishing_ grammar, assuming that things like this landed back in the xml.

**Comment by reschke** [comment]
> What Julian is missing in my argument is that if we change something like how we generate ToC, we still want to know what was generated at the time the RFC was published. (No, I don't think that we would necessarily re-render older RFCs with a newer way of ToC-ing should this ever happen.) This keeps truth in the right place.

I have trouble seeing why that would be useful. Can you come up with a specific example? (I can see *fixes* to the ToC generation, but in that case, I would hope these fixes are applied to all RFCs, not just new ones).

> I would be fine if there were a way to say things like this that didn't involve bleeding into the authoring grammar, but that would require a different publishing grammar, assuming that things like this landed back in the xml.

If you absolutely want these things in the XML, then they need to be distiguishable from what the author wrote. That actually os the case for \boilerplate> and \toc>, so tools can easily ignore those. It is not the case for the index and the author's section, and I would consider that a bug. (see https://github.com/reschke/xml2rfc/commit/e9396a562a79aec0dc8e0fcb6fa22aa2094db97b)

IMHO the right approach would have been to kick those elements into an extension namespace (and not include them in the published grammar); that makes it clear that they do not belong to the author grammar.

**Comment by jrlevine** [comment]
I'm inclined to agree with the sentiment that the prep tool's actions should be documented and deterministic. "Whatever xml2rfc did in 2021" is awfully squishy.

**Comment by JayDaley** [comment]
Here's an attempt to summarise the issues:

1. Do we need to know what was generated by the prep tool in published RFCs?
My view is no, we don't but more on this below.  There are all sorts of paths/processes that contribute to the final document and once the document is published none of that matters, it's just XML

2.  How does knowing what was generated by the prep tool relate to re-rendering of RFCs?
My view is that this is entirely orthogonal to re-rendering of RFCs and discussing re-rendering is simply complicating this discussion.

3.  Do we need to keep the prep tool grammar separate from the authoring grammar?
Yes absolutely because not doing this just confuses authors and makes the documentation hard to parse. 

4.  How do we keep the prep tool grammar separate from the authoring grammar?
We have identified two ways of doing this - having two grammars or using an extension namespace.  Rather than list the pros/cons with each one, I'll just say that I also prefer an extension namespace as that is easier for authors and shifts the "cost" of processing the document to the readers XML tools rather than the readers brain.

**Comment by jrlevine** [comment]
We're not disagreeing. The prep part needs to be well defined so it can be deterministically updated or removed.

**Comment by JayDaley** [comment]
> it's supposed to be possible to un-prep a document so you can get back an editable version.
Are there any use cases for this?  Are we expecting to lose the unprepared versions?

**Comment by jrlevine** [comment]
It doesn't seem like a great idea to depend on the RPC always having the unprepped versions forever.
There's also the question of whether it makes sense to prep or unprep a draft, or only an RFC ready for publication.

**Comment by JayDaley** [comment]
@jrlevine @cabo @rjsparks @reschke Just bumping this - If we can agree that unprep is not a necessary operation for us to support then I think it will save a lot of cycles.

**Comment by reschke** [comment]
I'm not sure *how* that would save cycles given what this ticket is about.

An rfcxml processor needs to be able to distinguish things added in the prep phase from the actually authored contents.

**Comment by cabo** [comment]
I would want to have an unprep if only to be able to write a test that the prep step preserves its input.
(And it should be trivial to write, commensurate with the effort you would spend on such a test.)

**Comment by cabo** [comment]
I'm not sure whether the participants of this discussion are aware that unprep is implemented:

https://xml2rfc.tools.ietf.org/xml2rfc-doc.html#appendix-B.3

``` 
--unprep
Reduce prepped xml to unprepped.
``` 

It's not perfect, as it leaves in attributes such as derivedCounter, but I think these are small bugs that can be easily fixed.

**Comment by JayDaley** [comment]
> I'm not sure whether the participants of this discussion are aware that unprep is implemented:

Yes that's my point - we have to use cycles maintaining it, which I worry will be wasted cycles because I don't see that any of the proposed use cases are sufficient to maintain it.

**Comment by cabo** [comment]
The feature is 135 LOC.  This serves

-- as documentation for the things that are final-form-only in publishing-RFCXML
-- as the basis for a test (that may still have to be written)
-- as a general tool for unprepping a published RFC (should there be a problem with the prerelease items)

Worth it for each of these.

**Comment by jrlevine** [comment]
That's largely true but prep also fills in defaults. I can see how an unprepper would go through and remove all of the attributes that have default values but ugh.

**Comment by cabo** [comment]
If the prep tool is filling in defaults (and ist most emphatically does, even the deprecated ones), the unprep needs to take them away again...

**Comment by reschke** [comment]
...where the trouble is that it won't know whether the default was added by the preptool or by the author :-)

**Comment by rjsparks** [comment]
(adding a bit):
If someone wants to pick up all or parts of expired draft X for which the author's source is lost or unavailable for one of many potential reasons, and that has its references expanded in the archive, it needs to be easy to generate a version with the references reverted to includes.
